plugin-server/tests/shared/process-event.ts

```diff
@@ -2157,5 +2192,143 @@ export const createProcessEventTests = (
         })
     }
 
+    test('set and set_once on the same key', async () => {
+        await createPerson(hub, team, ['distinct_id1'])
+
+        await processEvent(
+            'distinct_id1',
+            '',
+            '',
+            {
+                event: 'some_event',
+                properties: {
+                    token: team.api_token,
+                    distinct_id: 'distinct_id1',
+                    $set: { a_prop: 'test-set' },
+                    $set_once: { a_prop: 'test-set_once' },
+                },
+            } as any as PluginEvent,
+            team.id,
+            now,
+            now,
+            new UUIDT().toString()
+        )
+        expect((await hub.db.fetchEvents()).length).toBe(1)
+
+        const [event] = await hub.db.fetchEvents()
+        expect(event.properties['$set']).toEqual({ a_prop: 'test-set' })
+        expect(event.properties['$set_once']).toEqual({ a_prop: 'test-set_once' })
+
+        const [person] = await hub.db.fetchPersons()
+        expect(await hub.db.fetchDistinctIdValues(person)).toEqual(['distinct_id1'])
+        expect(person.properties).toEqual({ a_prop: 'test-set' })
+    })
+
+    describe('ingestion in any order', () => {
+        const ts0: DateTime = now
+        const ts1: DateTime = now.plus({ minutes: 1 })
+        const ts2: DateTime = now.plus({ minutes: 2 })
+        const ts3: DateTime = now.plus({ minutes: 3 })
+        // key encodes when the value is updated, e.g. s0 means only set call for the 0th event
+        // s03o23 means via a set in events number 0 and 3 plus via set_once on 2nd and 3rd event
+        // the value corresponds to which call updated it + random letter (same letter for the same key)
+        // the letter is for verifying we update the right key only
+        const set0: Properties = { s0123o0123: 's0a', s02o13: 's0b', s013: 's0e' }
+        const setOnce0: Properties = { s0123o0123: 'o0a', s13o02: 'o0g', o023: 'o0f' }
+        const set1: Properties = { s0123o0123: 's1a', s13o02: 's1g', s1: 's1c', s013: 's1e' }
+        const setOnce1: Properties = { s0123o0123: 'o1a', s02o13: 'o1b', o1: 'o1d' }
+        const set2: Properties = { s0123o0123: 's2a', s02o13: 's2b' }
+        const setOnce2: Properties = { s0123o0123: 'o2a', s13o02: 'o2g', o023: 'o2f' }
+        const set3: Properties = { s0123o0123: 's3a', s13o02: 's3g', s013: 's3e' }
+        const setOnce3: Properties = { s0123o0123: 'o3a', s02o13: 'o3b', o023: 'o3f' }
```

Comment:
I'll merge it, but please clean up this stuff in a follow up PR. This is not easy-to-read maintainable code

Reply:
How do you think this should be done, so we have the same coverage and it's easy to extend?
I was thinking that if we find a bug we can just add some more set/set_once values here to cover more cases.